src/components/toolbar/helpers.ts

```diff
@@ -207,41 +212,70 @@ export const promoteParagraphToSection = (
     offset += precedingSection.nodeSize
   }
 
-  const textContent = paragraph.textContent
+  const textContent = findChildrenByType(paragraph, schema.nodes.text).map(
+    ({ node }) => node
+  )
 
   const sectionTitle: SectionTitleNode = textContent
-    ? nodes.section_title.create({}, schema.text(textContent))
+    ? nodes.section_title.create({}, Fragment.from(textContent))
     : nodes.section_title.create()
 
   let sectionContent = Fragment.from(sectionTitle)
   if (endIndex < parentSection.childCount) {
     sectionContent = sectionContent.append(
       parentSection.content.cut(afterParagraphOffset)
     )
-  } else {
-    sectionContent = sectionContent.append(Fragment.from(paragraph.copy()))
   }
 
   if (parentSection.type.name === 'body') {
-    const newSection = nodes.section.create({}, sectionContent)
-    items[0] = nodes.body.create(
-      {},
-      items[0]
-        ? items[0].content.append(Fragment.from(newSection))
-        : Fragment.from(newSection)
-    )
+    const resolvePos = findChildrenByType(parentSection, nodes.section)[0]
+    if (resolvePos) {
+      const { pos } = resolvePos
+      items[0] = nodes.section.create(
+        {},
+        Fragment.from(sectionTitle).append(
+          parentSection.slice(afterParagraphOffset, pos).content
+        )
+      )
+      afterParentSection = beforeParentSection + pos
+      beforeParentSection = beforeParagraph
+    } else {
+      items[0] = nodes.section.create(
+        {},
+        Fragment.from(sectionTitle).append(
+          parentSection.slice(afterParagraphOffset).content
+        )
+      )
+    }
+
+    if ($beforeParagraph.index() === 0) {
+      if (!resolvePos) {
+        beforeParentSection = beforeParentSection + 1
+        afterParentSection = afterParentSection - 1
+      }
+    } else {
+      if ($beforeParagraph.nodeBefore) {
```

Comment:
please store this in `const nodeBefore = $beforeParagraph.nodeBefore` to be re used.